Add support for quantized operator conversion from PT to C2 via ONNX

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29694 Add support for quantized operator conversion from PT to C2 via ONNX**

Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear and Conv operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

supriyar committed:
Add support for quantized operator conversion from PT to C2 via ONNX
Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

houseroad commented on caffe2/onnx/backend.cc:
```diff
@@ -1684,9 +1684,17 @@ template <typename T>
-      onnx_tensor.data_type() == TensorProto::BOOL ? "GivenTensorBoolFill"
-                                                   : "GivenTensorIntFill");
+  std::string c2_op_type;
+  if (onnx_tensor.data_type() == TensorProto::BOOL) {
```
Nit: use switch instead of if-else if-else

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -0,0 +1,262 @@
+double getScaleFromInput(Node* input_node) {
+  c10::optional<IValue> scale;
+  std::string input_name = input_node->kind().toQualString();
+  if (input_name.find("quantize_per_tensor") != std::string::npos) {
```
you may want a switch statement on input_name

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -0,0 +1,262 @@
+    std::vector<int64_t> shapes,
+    double scale,
+    int64_t zero_point) {
+  Node* cast_node = graph->create(Symbol::caffe2("Int8GivenTensorFill"));
```
constant_node sounds more appropriate.

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -0,0 +1,262 @@
+    std::vector<int64_t> shapes,
+    double scale,
+    int64_t zero_point) {
+  Node* cast_node = graph->create(Symbol::caffe2("Int8GivenIntTensorFill"));
```
constant_node instead of cast_node?

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -0,0 +1,262 @@
+
+    // Permute weights?
+    /*
+    if (unpacked_weight.ndimension() == 2) {
```
Do we need permute?

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -0,0 +1,262 @@
+    qlinear_node->removeInput(1);
+
+    // Convert from int8 to uint8
+    int8_t* inp_data = (int8_t*)unpacked_weight.data_ptr<c10::qint8>();
```
i am a bit confused here. c10::qint8 is uint8? why do we need to do the conversion here? can we move the conversion to the actually assignment?

supriyar replied:
Pytorch uses int8 for the weights. But caffe2 (qnnpack) expects uint8 values. Hence the conversion. We do the same for quantization on mobile as well where we use qnnpack.

houseroad replied:
my question is what's the relationship between qint8 and int8_t?

Nit: reinterpret_cast<int8*>

supriyar replied:
c10::qint8 is actually int8. I need to do this cast here so that I can convert to uint8 by adding 128 below and casting to char.

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -0,0 +1,262 @@
+
+    // Convert from int8 to uint8
+    int8_t* inp_data = (int8_t*)unpacked_weight.data_ptr<c10::qint8>();
+    auto weight_zp = unpacked_weight.q_zero_point() + 128;
```
could you give the concrete type instead of using auto here?

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -0,0 +1,262 @@
+    // Create caffe2::Int8GivenTensorFill node
+    std::string w_data;
+    for (int64_t i = 0; i < wt_numel; ++i) {
+      w_data += static_cast<char>(inp_data[i] + 128);
```
Use std::ostringstream instead of += operator to be more efficient

houseroad commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -46,9 +45,31 @@ def pixel_shuffle(g, self, upscale_factor):
         return _unimplemented("pixel_shuffle", "only support 4d input")
     return g.op("DepthToSpace", self, blocksize_i=upscale_factor, mode_s="CRD")
 
+def interpolate_quantized(g, input, output_size, dim, interpolate_mode, align_corners=None):
```
Seems we don't have any symbolic to cover this yet.

houseroad commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -46,9 +45,31 @@ def pixel_shuffle(g, self, upscale_factor):
 
 def _interpolate(name, dim, interpolate_mode):
     def symbolic_fn(g, input, output_size, align_corners=None):
+        if input in sym_help._quantized_ops:
```
I would suggest not touching opset 11 in this PR. We can use a separate PR to do it.

supriyar replied:
I can remove this change, it was for a specific model

houseroad commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -447,6 +448,9 @@ def prelu(g, self, weight):
 
 
 def relu(g, input):
+    if input in sym_help._quantized_ops:
```
My suggestion is that, let's override the relu symbolic in symbolic_caffe2.py. It will make our change non-intrusive to the regular symbolics.

supriyar replied:
When we process the symbolics, we process the aten::relu function and arrive here first. quantized relu does not have a `quantized` prefix in the name, so there is no way to know if it is quantized or not without checking the input.

houseroad commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -2087,3 +2090,16 @@ def _weight_norm(g, weight_v, weight_g, dim):
         return g.op("ATen", weight_v, weight_g, dim_i=dim, operator_s="_weight_norm")
+
+# Ops below are for PyTorch Quantization conversion process.
+@parse_args('v', 'f', 'i', 't')
```
ditto, let's define this symbolic in symbolic_caffe2.py, and register it there.

supriyar replied:
Same reason as relu for why I couldn't do this. These ops below are in the `aten` namespace

houseroad commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -2087,3 +2090,16 @@ def _weight_norm(g, weight_v, weight_g, dim):
+def quantize_per_tensor(g, input, scale, zero_point, dtype):
+    return sym_caffe2.quantize_per_tensor(g, input, scale, zero_point)
+
+@parse_args('v')
```
ditto

houseroad commented on torch/onnx/symbolic_registry.py:
```diff
@@ -22,6 +22,17 @@ def register_version(domain, version):
         _registry[(domain, version)] = {}
     register_ops_in_version(domain, version)
 
+def register_quantized_ops(domain, version):
```
let's also move this function to symbolic_caffe2.py as well. Right now, it's very specific to caffe2.

houseroad commented on torch/onnx/utils.py:
```diff
@@ -611,6 +618,8 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
         import torch.onnx.symbolic_registry as sym_registry
 
+        if operator_export_type == OperatorExportTypes.ONNX_ATEN_FALLBACK:
+            sym_registry.register_quantized_ops('caffe2', opset_version)
```
you can change it to register_c2_ops, and register all ops in symoblics_caffe2, including overriding some regular ops' symbolics, such as relu, to support the quantized version.

houseroad reviewed (changes requested):
Great progress, left a few comments. Please address them.

houseroad commented:
The CI is broken too, check: https://circleci.com/gh/pytorch/pytorch/3580194?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

dreiss commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -46,9 +45,31 @@ def pixel_shuffle(g, self, upscale_factor):
+    output_size = g.op("Cast", output_size, to_i=sym_help.cast_pytorch_to_onnx["Long"])
+    output_size = g.op("Concat", input_size, output_size, axis_i=0)
+
+    output = g.op("Resize",
```
I don't understand.  If this is interpolate_quantized, how come we're not using Int8ResizeNearest?

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -2438,5 +2439,84 @@ def setup_rnn_tests():
+        q_model = torch.quantization.prepare(model, inplace=False)
+        q_model = torch.quantization.convert(q_model, inplace=False)
+        pytorch_res = q_model(*pt_inputs)
+        torch.onnx.export(q_model, pt_inputs, "model.onnx", verbose=True, input_names=input_names, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```
try io.BytesIO, instead of real files.

houseroad commented on torch/onnx/symbolic_helper.py:
```diff
@@ -513,3 +513,5 @@ def _cast_func_template(to_i, g, input, non_blocking):
     cast_pytorch_to_onnx["Bool"],
 ]
+
+_quantized_ops = set()
```
add some comments about this set?

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -0,0 +1,244 @@
+    auto weight_scale = unpacked_weight.q_scale();
+
+    auto input_val = match_vmap.at(vmap.at("r"))->node()->inputs()[0];
+    TORCH_INTERNAL_ASSERT(input_val->type()->isSubtypeOf(TensorType::get()));
```
print out the type if assertion fails?

houseroad reviewed (commented):
Thanks for address old comments, a few more comments.

houseroad commented:
Also please make sure the CI passing.

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

houseroad commented:
Also could you rebase to master to avoid /Build Failed CI, it will keep sending us emails.

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -17,6 +17,7 @@
 from torch.nn.utils import rnn as rnn_utils
 from debug_embed_params import run_embed_params
 import io
+import os
```
remove it?

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -2438,5 +2438,89 @@ def setup_rnn_tests():
                                            embed_params=True, opset_version=10))
 
 
+class TestQuantizedOps(unittest.TestCase):
```
Actually, why not create a test_pytorch_onnx_quantization_caffe2.py? I didn't see you reuse anything in this file.

supriyar replied:
Done

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear and Conv operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

dreiss commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -0,0 +1,246 @@
+    auto tmp = input_node->inputs()[0]->node();
+    return getScaleFromInput(tmp);
+  }
+  return 1.0;
```
Do we need a default value here?  Why not throw an exception?  Wouldn't that be safer?

supriyar replied:
Yes, that makes sense. I'll update it.

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear and Conv operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear and Conv operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

houseroad commented on test/onnx/test_pytorch_onnx_caffe2_quantized.py:
```diff
@@ -0,0 +1,131 @@
+        self.generic_test(QAddModule(), (x, y), input_names=["x", "y"])
+
+    def test_quantized_relu(self):
+        self.generic_unary_test(torch.nn.ReLU())
```
Nit: also add test to make sure regular relu can be correctly handled as well?

supriyar replied:
`test_linear` in test_pytorch_onnx_caffe2.py already covers regular relu.

houseroad commented on test/onnx/test_pytorch_onnx_caffe2_quantized.py:
```diff
@@ -0,0 +1,131 @@
+
+        model = torch.jit.load(buf)
+        input_names = ["x"]
+        f = io.BytesIO()
```
Nit: extract this export flow to a shared function?

houseroad reviewed (approved):
Looks great! left two more comments. Also please import the PR, and make sure internal signals are good, then we are good to go! Cheers :-)

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name '_caffe2' to register the ops if they are in the "quantized" namespace.
The change also adds JIT pass to unpack the quantized weights and insert the unpacked values into the graph.
The actual tensor values are looked up from the params dict.

Test Plan:
Currently tested end-to-end for Linear and Conv operator

python test/onnx/test_pytorch_onnx_caffe2.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18467130](https://our.internmc.facebook.com/intern/diff/D18467130)

[ghstack-poisoned]

yangboqwer commented:
@houseroad @supriyar    Hi, I have a problem when dump my quantized pytorch model to onnx. 

Traceback (most recent call last):
  File "get_pck_quant.py", line 129, in <module>
    onnx_model = export_to_onnx(model, x, input_names)
  File "get_pck_quant.py", line 118, in export_to_onnx
    operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
  File "/home/alohali/.local/lib/python3.5/site-packages/torch/onnx/__init__.py", line 148, in export
    strip_doc_string, dynamic_axes, keep_initializers_as_inputs)
  File "/home/alohali/.local/lib/python3.5/site-packages/torch/onnx/utils.py", line 66, in export
    dynamic_axes=dynamic_axes, keep_initializers_as_inputs=keep_initializers_as_inputs)
  File "/home/alohali/.local/lib/python3.5/site-packages/torch/onnx/utils.py", line 416, in _export
    fixed_batch_size=fixed_batch_size)
  File "/home/alohali/.local/lib/python3.5/site-packages/torch/onnx/utils.py", line 296, in _model_to_graph
    fixed_batch_size=fixed_batch_size, params_dict=params_dict)
  File "/home/alohali/.local/lib/python3.5/site-packages/torch/onnx/utils.py", line 130, in _optimize_graph
    torch._C._jit_pass_onnx_unpack_quantized_weights(graph, params_dict)
RuntimeError: false INTERNAL ASSERT FAILED at /pytorch/torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:86, please report a bug to PyTorch. Unrecognized quantized operator while trying to compute q_scale for operator aten::max_pool2d (getScaleFromInput at /pytorch/torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:86)

So, how to deal with aten::max_pool2d?

supriyar commented:
Are you trying to convert the model to Caffe2? Currently this conversion flow only supports conversion to C2.

Your max_pool2d operator doesn't appear to be quantized. If it is quantized the name should be `quantized::max_pool2d` https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qpool.cpp#L414

To add similar support for quantized::max_pool2d you can follow the steps we took for aten::avg_pool2d
https://github.com/pytorch/pytorch/blob/master/torch/onnx/symbolic_caffe2.py#L159

boydyang1 commented:
Thank you for your reply!
I'm trying to convert the model to C2
Here is my code(#33932), I don't know why my max_pool2d operator doesn't appear to be quantized.
Is there something wrong in my covert script？

supriyar commented:
Answered in #33932
